relay-metrics/src/aggregator/inner.rs

```diff
@@ -0,0 +1,926 @@
+    ///
+    /// The total length of the ring-buffer is then determined by the amount of time slots times
+    /// the amount of partition. In other words, every time slot as [`Self::num_partitions`]
+    /// partitions.
+    ///
+    /// Layout:
+    /// Time slots: [            ][            ][            ]
+    /// Partitions:  [    ][    ]  [    ][    ]  [    ][    ]
+    ///
+    /// An item is assigned by first determining it's time slot and then assigning it a partition
+    /// based on selected [`Self::partition_by`] strategy.
+    ///
+    /// The first item in the buffer is tracked by [`Self::head`] which is at any time the
+    /// current partition since the beginning "zero". The beginning in the aggregator refers to the
+    /// beginning of the epoch. The [`Self::head`] is calculated with `f(x) = x / bucket_interval * num_partitions`.
+    ///
+    /// Flushing a partition advances the [`Self::head`] by a single value `+1`. Meaning
+    /// effectively time is advanced by `bucket_interval / num_partitions`.
+    slots: VecDeque<Slot>,
+    /// The amount of partitions per time slot.
+    num_partitions: u32,
+
+    /// Position of the first element in [`Self::slots`].
+    head: u64,
+
+    /// Size of each individual bucket, inputs are truncated to this value.
+    bucket_interval: u32,
+    /// Amount of slots which is added to a bucket as a delay.
+    ///
+    /// This is a fixed delay which is added to to the time returned by [`Self::next_flush`].
+    delay: u32,
+
+    /// Total stats of the aggregator.
+    stats: stats::Total,
+    /// Configured limits based on aggregator stats.
+    limits: stats::Limits,
+
+    /// The maximum amount of slots (size of a `bucket_interval`) the timestamp is allowed to be
+    /// in the past or future.
+    slot_diff: SlotDiff,
+
+    /// Determines how partitions are assigned based in the input bucket.
+    partition_by: FlushBatching,
+    /// Hasher used to calculate partitions.
+    hasher: ahash::RandomState,
+}
+
+impl Inner {
+    pub fn new(config: Config) -> Self {
+        let bucket_interval = config.bucket_interval.max(1);
+        // Extend the configured time slots with the delay (in slots), to make sure there is
+        // enough space to satisfy the delay.
+        //
+        // We cannot just reserve space for enough partitions to satisfy the delay, because we have
+        // no control over which partition a bucket is assigned to, so we have to prepare for the
+        // 'worst' case, and that is full extra time slots.
+        let num_time_slots = config.num_time_slots.max(1) + config.delay.div_ceil(bucket_interval);
+        let num_partitions = config.num_partitions.max(1);
+
+        let mut slots = Vec::new();
+        for _ in 0..num_time_slots {
+            for partition_key in 0..num_partitions {
+                slots.push(Slot {
+                    partition_key,
+                    ..Default::default()
+                });
+            }
+        }
```

Comment:
nit: In combination with `VecDeque::from_iter`
```suggestion
        let slots = (0..num_time_slots).flat_map(|_| {
            (0..num_partitions).map(|partition_key| Slot {
                partition_key,
                ..Default::default()
            })
        });
```

You should be able to get `slots.len()` from the VecDeque, right?

Reply:
Do you think this is more readable? For me the nested for is more clear.

Performance wise it is the same, the conversion from `Vec` to `VecDeque` is "free", in fact the `FromIterator` impl collects into a `Vec` first.

Reply:
The procedural version is fine with me too.